packages/runtime/container-runtime/src/pendingStateManager.ts

```diff
@@ -165,12 +209,82 @@ export class PendingStateManager {
         this.pendingStates.push(pendingFlush);
     }
 
+    public rebaseAt(seqNum: number) {
+        const rebasePromises: Promise<void>[] = [];
+        // rebase initial ops at sequence number
+        while (!this.initialStates.isEmpty()) {
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            const nextState = this.initialStates.peekFront()!;
+            if (nextState.type === "message") {
+                if (nextState.referenceSequenceNumber > seqNum) {
+                    break; // nothing left to do at this sequence number
+                } else if (nextState.referenceSequenceNumber > 0 && nextState.referenceSequenceNumber < seqNum) {
```

Comment:
I'd not check for > 0. You can just assert somewhere (where you peak) that it's positive number.

Reply:
We need to correctly handle referenceSequenceNumber === 0 (the only valid case where we won't see the reference sequence number of a pending op)

Reply:
Can you please clarify the scenario? I believe that if file is created from snapshot (not from ops), then every pending op has referenceSequenceNumber that points to existing (sequenced) op, meaning it's positive. What is the case where we do not set referenceSequenceNumber ?

Reply:
We get pending ops with referenceSequenceNumber === 0 in tests, I'm not sure what would guarantee it should be positive elsewhere.

Reply:
I believe this is because our tests create empty files (i.e. do not use detached container flow). That's probably fine, but we really need to deprecate that. @curtisman - FYI, as we chatted about it (and impact it has in Container.ts around default connection modes). Also +@agarwal-navin as he did some overall cleanup in this space, it would be great to get assessment of how far we are from deprecating legacy workflow.